services/worker-manager/src/providers/azure.js

```diff
@@ -0,0 +1,379 @@
+
+        virtualMachine = await this._enqueue('query', () => this.computeClient.virtualMachines.createOrUpdate(resourceGroupName, virtualMachineName, {
+          ...cfg,
+          osProfile: {
+            adminUsername: 'worker',
+            ...cfg.osProfile,
+            // throw this away
+            adminPassword: slugid.nice(),
+            computerName: virtualMachineName,
+            customData,
+          },
+          networkProfile: {
+            ...cfg.networkProfile,
+            networkInterfaces: [
+              {
+                id: networkInterface.id,
+                primary: true,
+              },
+            ],
+          },
+          priority: cfg.priority || "Spot",
+          evictionPolicy: cfg.evictionPolicy || "Deallocate",
+          billingProfile: {
+            "maxPrice": -1,
+            ...cfg.billingProfile,
+          },
+          tags: {
+            ...cfg.tags || {},
+            'created-by': `taskcluster-wm-${this.providerId}`.replace(/[^a-zA-Z0-9-]/g, '-'),
+            'managed-by': 'taskcluster',
+            'worker-pool-id': workerPoolId.replace(/[^a-zA-Z0-9-]/g, '-').toLowerCase(),
+            'owner': workerPool.owner.replace(/[^a-zA-Z0-9-]/g, '-').toLowerCase(),
+          },
+        }));
+      } catch (err) {
+        await this._removeWorker({...providerData});
+        await workerPool.reportError({
+          kind: 'creation-error',
+          title: 'VM Creation Error',
+          description: err.message,
+          extra: err.details,
+        });
+        return;
+      }
+      this.monitor.log.workerRequested({
+        workerPoolId,
+        providerId: this.providerId,
+        workerGroup: this.providerId,
+        workerId: virtualMachine.vmId,
+      });
+      const now = new Date();
+      await this.Worker.create({
+        workerPoolId,
+        providerId: this.providerId,
+        workerGroup: this.providerId,
+        workerId: virtualMachine.vmId,
+        created: now,
+        lastModified: now,
+        lastChecked: now,
+        expires: taskcluster.fromNow('1 week'),
+        state: this.Worker.states.REQUESTED,
+        capacity: cfg.capacityPerInstance,
+        providerData: {
+          ...providerData,
+          registrationExpiry,
+        },
+      });
+    }));
+  }
+
+  async deprovision({workerPool}) {
+    // nothing to do: we just wait for workers to terminate themselves
+  }
+
+  async registerWorker({worker, workerPool, workerIdentityProof}) {
+    this.monitor.log.workerRunning({
+      workerPoolId: workerPool.workerPoolId,
+      providerId: this.providerId,
+      workerId: worker.workerId,
+    });
+    await worker.modify(w => {
+      w.lastModified = new Date();
+      w.state = this.Worker.states.RUNNING;
+    });
+    // assume for the moment that workers self-terminate before 96 hours
+    return {expires: taskcluster.fromNow('96 hours')};
+  }
+
+  async scanPrepare() {
+    this.seen = {};
+    this.errors = {};
```

Comment:
I think that `this.errors` is functionally unused in this provider. Let's rip it out from here and its other uses below. It is just used in the google provider to get reports from operations.